pkg/awsutils/awsutils.go

```diff
@@ -513,75 +526,246 @@ func (cache *EC2InstanceMetadataCache) initWithEC2Metadata(ctx context.Context)
+
+	return filteredENIs.SortedList()
+}
 
+// Helper function to apply security groups to a list of ENIs
+func (cache *EC2InstanceMetadataCache) applySecurityGroupsToENIs(ctx context.Context, eniIDs []string, sgIDs []string, logPrefix string) {
+	for _, eniID := range eniIDs {
+		log.Debugf("%s ENI %s with security groups %v", logPrefix, eniID, sgIDs)
+
+		attributeInput := &ec2.ModifyNetworkInterfaceAttributeInput{
+			Groups:             sgIDs,
+			NetworkInterfaceId: aws.String(eniID),
+		}
+		start := time.Now()
+		_, err := cache.ec2SVC.ModifyNetworkInterfaceAttribute(ctx, attributeInput)
+		prometheusmetrics.Ec2ApiReq.WithLabelValues("ModifyNetworkInterfaceAttribute").Inc()
+		prometheusmetrics.AwsAPILatency.WithLabelValues("ModifyNetworkInterfaceAttribute", fmt.Sprint(err != nil), awsReqStatus(err)).Observe(msSince(start))
+
+		if err != nil {
+			if errors.As(err, &awsAPIError) {
+				if awsAPIError.ErrorCode() == "InvalidNetworkInterfaceID.NotFound" {
+					awsAPIErrInc("IMDSMetaDataOutOfSync", err)
+				}
+			}
+			checkAPIErrorAndBroadcastEvent(err, "ec2:ModifyNetworkInterfaceAttribute")
+			awsAPIErrInc("ModifyNetworkInterfaceAttribute", err)
+			prometheusmetrics.Ec2ApiErr.WithLabelValues("ModifyNetworkInterfaceAttribute").Inc()
+			log.Warnf("%s: unable to update ENI %s security groups: %v", logPrefix, eniID, err)
+		}
+	}
+}
+
+// Helper function to detect and log security group changes
+func (cache *EC2InstanceMetadataCache) detectSecurityGroupChanges(newSGs []string, currentSGs *StringSet, sgType string) (int, int) {
+	newSGSet := StringSet{}
+	newSGSet.Set(newSGs)
+	addedSGs := newSGSet.Difference(currentSGs)
+	deletedSGs := currentSGs.Difference(&newSGSet)
+
+	addedCount := 0
 	for _, sg := range addedSGs.SortedList() {
-		log.Infof("Found %s, added to ipamd cache", sg)
-		addedSGsCount++
+		log.Infof("Found %s SG %s, added to ipamd cache", sgType, sg)
+		addedCount++
 	}
+
+	deletedCount := 0
 	for _, sg := range deletedSGs.SortedList() {
-		log.Infof("Removed %s from ipamd cache", sg)
-		deletedSGsCount++
+		log.Infof("Removed %s SG %s from ipamd cache", sgType, sg)
+		deletedCount++
 	}
-	cache.securityGroups.Set(sgIDs)
 
-	if !cache.useCustomNetworking && (addedSGsCount != 0 || deletedSGsCount != 0) {
-		var eniIDs []string
+	return addedCount, deletedCount
+}
+
+// RefreshCustomSGIDs discovers and refreshes security groups tagged for use with the CNI
+func (cache *EC2InstanceMetadataCache) RefreshCustomSGIDs(ctx context.Context, dsAccess *datastore.DataStoreAccess) error {
+	sgIDs, err := cache.discoverCustomSecurityGroups(ctx)
+	if err != nil {
+		awsAPIErrInc("DiscoverCustomSecurityGroups", err)
+		log.Warnf("Failed to discover custom security groups: %v", err)
+		log.Info("Falling back to using primary security groups for ENIs in secondary subnets")
+
+		// Clear custom security groups cache to trigger fallback behavior
+		cache.customSecurityGroups.Set([]string{})
```

Comment:
Not following why need to flush the cache here instead of handling them in fallback call below?